### PR TITLE
fix(pipeline): Fix readable state

### PR DESCRIPTION
**Pull request by jhermsmeier**
**Changes:**
- **lib:** Make the Pipeline a `class`, removing the need for the `bloodline` dependency
- **lib:** Reduce unnecessary complexity in `.splice()` through addition of `._beforeMutate()` and `._afterMutate()`
- **lib:** Fix not reading all available chunks in `._read()`, causing a stall when multiple chunks were available while waiting for the "readable" event
- **lib:** Fix `.end()` not always "ending" the pipeline itself
- **lib:** Fix not waiting for the writable state of the pipeline to finish before ending the readable state
- **lib:** Properly end the readable state by pushing `null` only when the writable has finished and all internal streams have ended (their write & read queues empty and flushed)
- **lib:** Fix not triggering internal write logic by using the proper `stream.write()` instead of `stream._write()` in `pipeline._write()`
- **test:** Only end tests on `end` instead of `finish`, waiting for the last readable in the pipeline to end (become empty)

Change-Type: patch

### CODE REVIEW

**Comment by jviotti** [comment]
Looks like node v4 is failing. Also, can you explain the issue/fix in the PR? Is this issue something we can capture on a test case?

File: lib/pipeline.js [hunk 1]
```diff
@@ -97,42 +151,37 @@ Pipeline.prototype = {
...
    * @returns {Boolean}
    */
-  _write( chunk, encoding, next ) {
+  _write(chunk, encoding, next) {
 
-    if( this._streams.length === 0 ) {
-      process.nextTick( next )
-      return this.push( chunk, encoding )
+    if( !this._streams.length ) {
+      this.push( chunk )
```
Comment (jhermsmeier) [inline lib/pipeline.js]: Looks like I forgot the `encoding` argument here, need to verify.

**Comment by jhermsmeier** [comment]
> Also, can you explain the issue/fix in the PR?

Yes, sorry, I meant to do that earlier, but forgot about it :/

> Is this issue something we can capture on a test case?

The minor changes to the tests should already catch this, but we should be able to capture it in more explicit detail. I'll see to adding another test.

**Review by jviotti — APPROVED** [review]

File: test/edge-cases.js [hunk 2]
```diff
@@ -14,10 +14,12 @@ describe( 'Edge cases', function() {
 
     pipeline
       .on( 'error', done )
-      .on( 'finish', done )
+      .on( 'end', done )
       .on( 'readable', function() {
         var hash = this.read()
-        assert.equal( hash.toString( 'hex' ), '37a309ae59ad3a562a083eb90e32fa6757bdd41f' )
+        if( hash != null ) {
```
Comment (lurch) [inline test/edge-cases.js:20]: So if the returned hash is always null, the test still passes??
Is this because the hash is null first time, but then set the second time, or something? Maybe it'd be worth checking that we see at least one non-null hash during the unit-test?
